### ISSUE TITLE
Web Setup Wizard 500 error when session storage is configured to use memcache

### ISSUE THREAD

**Issue by dotsoa** [magento/magento2#9633]
### Preconditions

1. Magento Version: 2.1.6
2. PHP-FPM - Specific pool is currently using  7.0.17
3. Apache 2.2.31

### Steps to reproduce

1. Perform a fresh installation of Magento 2.1.6
2. Switch and configure session storage method to memcache in env.php
3. Log into admin and navigate to System > Tools > Web Setup Wizard

### Expected result
1. The Setup Wizard page loading successfully

### Actual result

1. Page returns an HTTP ERROR 500. Additionally here is the error from the PHP logs when this occurs:

`[15-May-2017 07:22:26 UTC] PHP Fatal error:  Uncaught Exception: Warning: SessionHandler::read(): open(unix:///var/tmp/memcached.magetwot.216.mage2test.com_sessions.sock?persistent=1&weight=2&timeout=10&retry_interval=0/sess_cvud2fpo9dbmofdhld3fap3682, O_RDWR) failed: No such file or directory (2) in /chroot/home/magetwot/216.mage2test.com/html/vendor/magento/framework/Session/SaveHandler/Native.php on line 22 in /chroot/home/magetwot/216.mage2test.com/html/vendor/magento/framework/App/ErrorHandler.php:61
Stack trace:
#0 [internal function]: Magento\Framework\App\ErrorHandler->handler(2, 'SessionHandler:...', '/chroot/home/ma...', 22, Array)
#1 /chroot/home/magetwot/216.mage2test.com/html/vendor/magento/framework/Session/SaveHandler/Native.php(22): SessionHandler->read('cvud2fpo9dbmofd...')
#2 /chroot/home/magetwot/216.mage2test.com/html/vendor/magento/framework/Session/SaveHandler.php(93): Magento\Framework\Session\SaveHandler\Native->read('cvud2fpo9dbmofd...')
#3 [internal function]: Magento\Framework\Session\SaveHandler->read('cvud2fpo9dbmofd...')
#4 in /chroot/home/magetwot/216.mage2test.com/html/vendor/magento/framework/App/ErrorHandler.php on line 61`

`[15-May-2017 07:22:26 UTC] PHP Fatal error:  Uncaught Exception: Warning: SessionHandler::write(): open(unix:///var/tmp/memcached.magetwot.216.mage2test.com_sessions.sock?persistent=1&weight=2&timeout=10&retry_interval=0/sess_cvud2fpo9dbmofdhld3fap3682, O_RDWR) failed: No such file or directory (2) in /chroot/home/magetwot/216.mage2test.com/html/vendor/magento/framework/Session/SaveHandler.php on line 105 in /chroot/home/magetwot/216.mage2test.com/html/vendor/magento/framework/App/ErrorHandler.php:61
Stack trace:
#0 [internal function]: Magento\Framework\App\ErrorHandler->handler(2, 'SessionHandler:...', '/chroot/home/ma...', 105, Array)
#1 /chroot/home/magetwot/216.mage2test.com/html/vendor/magento/framework/Session/SaveHandler.php(105): SessionHandler->write('cvud2fpo9dbmofd...', '')
#2 [internal function]: Magento\Framework\Session\SaveHandler->write('cvud2fpo9dbmofd...', '')
#3 {main}
  thrown in /chroot/home/magetwot/216.mage2test.com/html/vendor/magento/framework/App/ErrorHandler.php on line 61`

As the title states. When the session storage method is set to memcache an 500 error happens when navigating to the Web Setup Wizard. With the session method set to files it loads as intended. Based off the error it seems like it's still trying to write/read the session as a file even though magento is configured to use the memcache socket. 

I've found as a workaround if i specifically set `php_value[session.save_handler] = memcache` in the PHP-FPM pool configuration file for this pool then the Setup Wizard page loads successfully. In the past this change has not been needed. Simply configuring memcache in env.php was enough.

**Comment by kozmikyak** [comment]
I can verify that the same thing happens (500 error on using the web wizard) with PHP 7.0.18 on Centos 7, and the latest install of Magento (from 2 days ago) in mod_php mode.

**Comment by digitalredefined** [comment]
I'm having the same issue as well on 2.1.7.

**Comment by TomashKhamlai** [comment]
Hello and thank you for your submission! Internal ticket MAGETWO-70986 was created to address this issue.

**Comment by magento-engcom-team** [comment]
@dotsoa, thank you for your report.
We've created internal ticket(s) MAGETWO-70986 to track progress on the issue.

**Comment by ghost** [comment]
Working on this for #squashtoberfest !

Though so far it seems that the setup wizard just never sets the session.save_handler value at all and doesn't include memcache as an option anywhere.

So what happens is it defaults to files but still uses the path and the tcp://... isn't a file.

**Comment by okorshenko** [comment]
Hi @sylink could you please accept the invitation to join magento2 repository on the GitHub?

**Comment by ghost** [comment]
No solution at the moment, we see the admin/setup and the frontend both set the save_handler differently and not explicitly? So a simple fix if you force the init_set but the code side appears to be a lot more complex than expected.

**Comment by ghost** [comment]
The issue is somewhere in the config where it gets the IniOptions in SessionManager.php , line 579

`   */
    private function initIniOptions()
    {
        foreach ($this->sessionConfig->getOptions() as $option => $value) {
            $result = ini_set($option, $value);
            if ($result === false) {
                $error = error_get_last();
                throw new \InvalidArgumentException(
                    sprintf('Failed to set ini option "%s" to value "%s". %s', $option, $value, $error['message'])
                );
            }
        }
`

A manual ini_set for memcache before the foreach allows setup to work with memcache, as in other locations, but only 7 values are set in that loop and none are session.save_handler

This leads me to Session/Config.php which gets these options but has no code to get the handler, so I think it needs added there ?

All in vendor/magento/framework/Session

**Comment by ghost** [comment]
Add this to constructor for vendor/magento/framework/Session/Config.php
, say line 137 or so:
```
$saveHandler=$deploymentConfig->get(self::PARAM_SESSION_SAVE_METHOD);
		if ($saveHandler) {
			$this->setOption('session.save_handler', $saveHandler);
		}
```

It looks like in that file this constant is set -

const PARAM_SESSION_SAVE_METHOD = 'session/save';

Then never utilized at all, so I added the block to grab the value from the deployment config. If that is blank it should default to files as expected.

**Comment by sivajik34** [comment]
checkout page is very slow.
magento 2.2.1
https://magento.stackexchange.com/questions/199973/magento-2-2-0-checkout-regenerate-session-id-very-slow

**Comment by okorshenko** [comment]
@dotsoa 
The issue has been fixed and delivered in 2.2-develop branch. Will be available with upcoming patch release.

**Comment by AlexisEvo** [comment]
@sivajik34 This is a separate problem which is unrelated to this issue. Please see #12742

**Comment by magento-engcom-team** [comment]
Hi @dotsoa. Thank you for your report. 
The issue has been fixed in magento-engcom/magento2ce#1296 by @magento-engcom-team in 2.3-develop branch
Related commit(s):
 - aefd089d3b29023d4ec09036f5ae836a97fe549b
 - 0c799d2defa8341b7eb23bd5c92717b5f64138e7
 - 0a282955d6ec5b8e8f8fc6cf94d1a813adebb285
 - d0d8d03982dde84fb119636492d3272b54bd59e0
 - 1623007dc26f1953b10893304f3d9b7efa8db805
 - 64f63b3cf56fe03ae78a0154133d27f05749be4a
 - eceb5eb5d31608ef96403dbd7857e8178acb5dc6
 - e1db353d9fd9ed508846160a3cdb16228c09b60a
 - 7b12d50d92e1503f959219a5c8f05ecbef1cc423
 - 60285c4011fa4ccabe6548ace2b6aefc0cc65c8f
 - 2d910df07653f1a3fac8fbf3688b57ce5e3869e7
 - 57c355a802a3e2ff21b7cf7312c203ab735cd2bd
 - 4f9f40e54207c31a2c5e8460c85a042ad39cfa66
 - ebd663b8dde32c49d9f147bd2bbe3b438a2f415e
 - bdb62bb6c6c8326b31c83fc4c6343ee4017b84e6
 - 76f6eed85e282be7df18766868f03a5e1aa4109c
 - cd5c72ee83a89fe3c6632ddf9b6aeabf512503fc
 - b03566074756b7264155e10e2d68845d5000df29
 - 6217c009f7565b72fe8bd24ca2e862131f3d8dee
 - e07d040a33324bae14961ccc629604738854684b
 - 47ea3188264a42f735d12c6c6ebe208135eefb69

The fix will be available with the upcoming patch release.
